pkg/gardenlet/controller/federatedseed/seed_control.go

```diff
@@ -0,0 +1,238 @@
+	"github.com/gardener/gardener/pkg/logger"
+	seedpkg "github.com/gardener/gardener/pkg/operation/seed"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// Controller is responsible for maintaining multiple federated Seeds' controllers
+type Controller struct {
+	federatedSeedControllerManager *federatedSeedControllerManager
+	k8sGardenClient                kubernetes.Interface
+
+	config *config.GardenletConfiguration
+
+	seedLister    gardencorelisters.SeedLister
+	seedHasSynced cache.InformerSynced
+	seedQueue     workqueue.RateLimitingInterface
+
+	recorder record.EventRecorder
+
+	ctx                    context.Context
+	numberOfRunningWorkers int
+	workerCh               chan int
+}
+
+// NewFederatedSeedController creates new controller that syncs extensions states to the ShootState resource in the garden cluster.
+func NewFederatedSeedController(ctx context.Context, k8sGardenClient kubernetes.Interface, gardenCoreInformerFactory gardencoreinformers.SharedInformerFactory, config *config.GardenletConfiguration, recorder record.EventRecorder) *Controller {
+	var seedInformer = gardenCoreInformerFactory.Core().V1alpha1().Seeds()
+
+	controller := &Controller{
+		k8sGardenClient: k8sGardenClient,
+		seedLister:      seedInformer.Lister(),
+		seedQueue:       workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "seed-controllers"),
+		config:          config,
+		ctx:             ctx,
+		workerCh:        make(chan int),
+		recorder:        recorder,
+	}
+
+	controller.federatedSeedControllerManager = &federatedSeedControllerManager{
+		seedShootStateSyncController: make(map[string]*shootstate.SyncController),
+	}
+
+	seedInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: controllerutils.SeedFilterFunc(confighelper.SeedNameFromSeedConfig(config.SeedConfig), config.SeedSelector),
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc:    controller.seedAdd,
+			UpdateFunc: controller.seedUpdate,
+			DeleteFunc: controller.seedDelete,
+		},
+	})
+
+	controller.seedHasSynced = seedInformer.Informer().HasSynced
+	return controller
+}
+
+func (c *Controller) seedAdd(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+
+		return
+	}
+	c.seedQueue.Add(key)
+}
+
+func (c *Controller) seedUpdate(oldObj, newObj interface{}) {
+	var (
+		newSeedObj = newObj.(*gardencorev1alpha1.Seed)
+		oldSeedObj = oldObj.(*gardencorev1alpha1.Seed)
+	)
+
+	newSeedBootstrappedCondition := helper.GetCondition(newSeedObj.Status.Conditions, gardencorev1alpha1.SeedBootstrapped)
+	oldSeedBootstrappedCondition := helper.GetCondition(oldSeedObj.Status.Conditions, gardencorev1alpha1.SeedBootstrapped)
+
+	if oldSeedBootstrappedCondition.Status != newSeedBootstrappedCondition.Status {
```

Comment:
Again, why do we look at the condition status here? I think it'd correct to only add the seed to the queue once  the bootstrapped condition was `True` once, but when it switches to `False` in the meantime we should not delete the seed from the queue. We should only delete the seed out of the queue if it has a deletion timestamp.

Reply:
Indeed, I agree with @rfranzke on this one as I mentioned previously.
This check should only do something (start the informers) if the SeedBootstrappedCondition changes from False to True. Otherwise, nothing should be done.

Reply:
Basically, we only want to add the seed on update if the condition has changed from some value (false or not initialised) to true. Nothing should be done if anything else changes or the condition status becomes false.